src/Link/Link.php

```diff
@@ -59,14 +58,15 @@ public function __construct($relation)
 
     /**
      * Factory for creating links
+     * $spec['url'] is deprecated since 1.5.0; use $spec['href'] instead
      *
-     * @param array $spec {
-     *      @var string|array $rel Required.
-     *      @var array $props Optional.
-     *      @var string $href Optional.
-     *      @var string|array $route Optional.
-     *      @var string $url {@deprecated since 1.5.0; use 'href' instead} Optional.
-     * }
+     * @psalm-param array{
```

Comment:
FYI, you can add a psalm type to the class-level doc block such as:

```php
/**
 * @psalm-type LinkArraySpec = array{
 *   rel: string|whatever
 *   other: list<whatever>
 * }
 */
```

And then use that type for your parameter and return docs: `@return LinkArraySpec` instead of copying and pasting the entire spec around.

In other classes, the type can be imported with `@psalm-import-type LinkArraySpec from ImportedClassName` on the class-level doc block.

Reply:
Hi @gsteel, I have tried to extract it to a type, but then psalm raises some `MixedArgumentTypeCoercion` errors due to the fact that the initial status of `$links` property at `src/Metadata/Metadata.php:88` is initialized as an empty array (read as `array<empty,empty>` by psalm)

Any change on the initial property status would be considered a BC, so I suggest keeping the duplicated `$spec` Psalm definition as added on the latest commit.

What do you think?

Reply:
You can make each of the keys optional:

```php
/**
 * @psalm-type LinkArraySpec = array{
 *     rel?: string|whatever,
 *     other?: list<whatever>
 * }
 */
```

Doing so gives you both the correct shape AND allows for empty arrays.